File: javascript/smartlist/smartlist-api-mysql-project/src/modules/report/report.candidate.controller.ts
import {
  Controller,
  Get,
  Param,
  HttpCode,
  ParseIntPipe,
  Res,
} from '@nestjs/common';
import { ApiOperation, ApiResponse } from '@nestjs/swagger';
import { ReportResponseCodes } from './report.response.codes';
import { HighLevelSummaryRespDto } from './dto/response/highlevel-summary.resp.dto';
import { Role } from 'src/enums/role';
import { Roles } from 'src/guards/roles.decorator';
import { ReportCandidateService } from './report.candidate.service';
import { CandidateAssessmentReportRespDto } from './dto/response/candidate-assessment-report.resp.dto';
import { Response } from 'express';
import { CandidateResponseCodes } from '../candidate/candidate.response.codes';
import { setDataAndHeaderInRes } from 'src/utils/file.utils';

@Controller('report/')
@Roles(Role.SUPERADMIN, Role.ADMIN, Role.MEMBER)
export class ReportCandidateController {
  constructor(
    private readonly reportCandidateService: ReportCandidateService,
  ) {}

  @ApiOperation({
    summary: 'Get candidate assessment report',
  })
  @ApiResponse({
    ...ReportResponseCodes.SUCCESS,
    type: [HighLevelSummaryRespDto],
  })
  @ApiResponse(ReportResponseCodes.BAD_REQUEST)
  @ApiResponse(ReportResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(ReportResponseCodes.INVALID_ASSESSMENT_ID)
  @HttpCode(200)
  @Get('candidate/:candidateId/assessment/:assessmentId')
  async getCandidateScoreForAssessmentBlocks(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('assessmentId', ParseIntPipe) assessmentId: number,
  ): Promise<CandidateAssessmentReportRespDto> {
    return await this.reportCandidateService.getCandidateAssessmentReport(
      candidateId,
      assessmentId,
    );
  }

  @ApiOperation({
    summary: 'Get candidate assessment report PDF',
  })
  @ApiResponse({
    ...ReportResponseCodes.SUCCESS,
  })
  @ApiResponse(ReportResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(ReportResponseCodes.INVALID_ASSESSMENT_ID)
  @ApiResponse(CandidateResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS)
  @HttpCode(200)
  @Get('candidate/:candidateId/assessment/:assessmentId/pdf')
  async downloadCandidateAssessmentReport(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('assessmentId', ParseIntPipe) assessmentId: number,
    @Res() res: Response,
  ): Promise<Response> {
    const buffer =
      await this.reportCandidateService.getCandidateAssessmentReportAsPDF(
        candidateId,
        assessmentId,
      );
    return setDataAndHeaderInRes(res, buffer, 'candidateassessment');
  }

  @ApiOperation({
    summary: 'Send candidate assessment report by mail',
  })
  @ApiResponse({
    ...ReportResponseCodes.SUCCESS,
  })
  @ApiResponse(ReportResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(ReportResponseCodes.INVALID_ASSESSMENT_ID)
  @ApiResponse(CandidateResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS)
  @HttpCode(200)
  @Get('candidate/:candidateId/assessment/:assessmentId/mail')
  async sendCandidateAssessmentReportMail(
    @Param('candidateId', ParseIntPipe) candidateId: number,
    @Param('assessmentId', ParseIntPipe) assessmentId: number,
  ): Promise<boolean> {
    return await this.reportCandidateService.sendCandidateAssessmentReportMail(
      candidateId,
      assessmentId,
    );
  }
}

File: java/smartlist/src/main/java/com/smartlist/module/report/ReportCandidateController.java
package com.smartlist.module.report;

import com.smartlist.module.report.dto.response.AssessmentStatusSummaryRespDTO;
import com.smartlist.module.report.dto.response.CandidateAssessmentReportRespDTO;
import com.smartlist.module.report.dto.response.HighLevelSummaryRespDTO;
import com.smartlist.module.report.dto.response.RateSummaryRespDTO;
import com.smartlist.utils.AppResponseCodes;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.constraints.Positive;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.annotation.Secured;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

@RestController
@Tag(name = "Report-Candidate", description = "Report-Candidate API")
@RequestMapping(value ="/report/candidate/", produces = { "application/json" })
@ResponseStatus(HttpStatus.OK)
@Validated
@Secured({"SUPERADMIN", "ADMIN", "MEMBER"})
public class ReportCandidateController {

    private final ReportCandidateService reportCandidateService;

    ReportCandidateController(final ReportCandidateService reportCandidateService){
        this.reportCandidateService = reportCandidateService;
    }

    @Operation(summary = "Get candidate assessment report")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = HighLevelSummaryRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("{candidateId}/assessment/{assessmentId}")
    public CandidateAssessmentReportRespDTO getCandidateScoreForAssessmentBlocks(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("assessmentId") @Positive Integer assessmentId){
        return this.reportCandidateService.getCandidateAssessmentReport(candidateId, assessmentId);
    }

    @Operation(summary = "Get candidate assessment report PDF")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = RateSummaryRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("{candidateId}/assessment/{assessmentId}/pdf")
    public ResponseEntity<byte[]> downloadCandidateAssessmentReport(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("assessmentId") @Positive Integer assessmentId) {
        return this.reportCandidateService.getCandidateAssessmentReportAsPDF(candidateId, assessmentId);
    }

    @Operation(summary = "Send candidate assessment report by mail")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = AssessmentStatusSummaryRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("{candidateId}/assessment/{assessmentId}/mail")
    public void getAssessmentStatusSummary(
            @PathVariable("candidateId") @Positive Integer candidateId,
            @PathVariable("assessmentId") @Positive Integer assessmentId) {
        this.reportCandidateService.sendCandidateAssessmentReportMail(candidateId, assessmentId);
    }





}
